osquery/tables/ipmi/ipmi_info.cpp

```diff
@@ -0,0 +1,1309 @@
+
+/// Unique pointer deleter for ipmi_lanparm_t*
+const auto kFreeLANParm = [](ipmi_lanparm_t* lp) {
+  int rv = ipmi_lanparm_destroy(lp, NULL, NULL);
+  if (rv != 0) {
+    LOG(WARNING) << "Did not successfully destroy ipmi_lanparm_t*: "
+                 << strerror(rv);
+  }
+};
+
+/**
+ * @brief retrieves name of MC
+ *
+ * @param mc pointer to OPENIPMI ipmi_mc_t
+ *
+ * @param defaultName fallback name if the MC name is unable to be retrieved
+ *
+ * @return MC name as std::string
+ */
+static inline std::string getMCName(ipmi_mc_t* mc,
+                                    std::string defaultName = "missing") {
+  char name[IPMI_MC_NAME_LEN];
+
+  int len = ipmi_mc_get_name(mc, name, IPMI_MC_NAME_LEN);
+  if (len < 1) {
+    return defaultName;
+  }
+
+  return name;
+}
+
+/**
+ * @brief generates key to uniquely identify an MC instance
+ *
+ * @param mc pointer to OPENIPMI ipmi_mc_t
+ *
+ * @return unique key for MC as std::string
+ */
+static inline std::string getMCKey(ipmi_mc_t* mc) {
+  return getMCName(mc) + "-" + std::to_string(ipmi_mc_device_id(mc));
+}
+
+class IPMIClient;
+
+/**
+ * @brief parmData encapsulates things that are required for retrieving LANPARM
+ *
+ * @param client pointer to IPMIClient instance
+ *
+ * @param mcKey unique key to identify the MC
+ *
+ * @parm colName the osquery table column name of the LANPARM
+ *
+ * @parm parm OpenIPMI identifier for the LANPARM
+ *
+ */
+struct parmData {
+  IPMIClient* client;
+  std::string mcKey;
+  std::string colName;
+  unsigned int parm;
+
+  parmData() : client(nullptr), mcKey(""), colName(""), parm(999) {}
+};
+
+/**
+ * @brief Client for interacting with OpenIPMI
+ *
+ * IPMIClient is a singleton object accessible only by the IPMIClient::get
+ * static method.  IPMIClient is lazily loaded on the first query to any IPMI
+ * table.  The initialization process waits for OpenIPMI to be in the "fully up"
+ * state so it can a few minutes.
+ */
+class IPMIClient : public InternalRunnable {
+ public:
+  /**
+   * @brief retrieves instance of IPMIClient (singleton).
+   *
+   */
+  static IPMIClient& get();
+
+ public:
+  /**
+   * @brief retrieves QueryData for ipmi_threshold_sensors.
+   *
+   * @param results Reference to QueryData.
+   */
+  void getThresholdSensors(QueryData& results);
+
+  /**
+   * @brief retrieves QueryData for ipmi_lan.
+   *
+   * @param results Reference to QueryData.
+   */
+  void getLANConfigs(QueryData& results);
+
+  /**
+   * @brief retrieves QueryData for ipmi_fru.
+   *
+   * @param results Reference to QueryData.
+   */
+  void getFRUs(QueryData& results);
+
+  /**
+   * @brief retrieves QueryData for ipmi_mc.
+   *
+   * @param results Reference to QueryData.
+   */
+  void getMCs(QueryData& results);
+
+  /**
+   * @brief checks if IPMIClient is successfully initiated.
+   *
+   * @return bool indicating client state.
+   */
+  bool up();
+
+  /// Starts background clean up routine.  Implements InternalRunnable.
+  void start() override;
```

Comment:
I believe start and stop are virtual methods that implement `InternalRunnable`...